### ISSUE TITLE
Training metrics logging + visualization

### ISSUE THREAD

**Issue by yash-jha-888** [yash-jha-888/aurelius#13]
<html>
<body>
<html><head></head><body><h1>Training metrics logging + visualization</h1>
<p><strong>Milestone:</strong> v0.3
<strong>Position in build order:</strong> after optimizer abstraction, <strong>before L2 regularization</strong></p>
<h2>Why this comes before regularization</h2>
<p>Right now the only training signal Aurelius emits is an epoch number and an accuracy printout. That is enough to know <em>whether</em> a run worked. It is not enough to know <em>why</em>.</p>
<p>Everything left in v0.3 is a decision that can only be made by looking at a curve:</p>
<ul>
<li><strong>L2 regularization</strong> is an intervention on the train/validation gap. Without both curves plotted, tuning lambda means guessing from a single final test number — with no way to distinguish "reduced overfitting," "damaged fit," and "changed nothing."</li>
<li><strong>LR scheduling</strong> is entirely about the <em>shape</em> of the descent. Shape is invisible in a terminal spew.</li>
<li><strong>Dropout</strong> has the same problem as L2.</li>
</ul>
<p>Building the instruments after the experiments that need them is backwards. This lands first.</p>
<h2>Scope</h2>
<h3>1. C++ side — a <code>Logger</code>, not scattered <code>cout</code></h3>
<p>A small class that owns the run's output. Design constraint, same principle as the activation refactor: <strong>the thing that prints to the terminal and the thing that writes the CSV must be the same call.</strong> If they're separate, they drift, and one day the plot disagrees with the log and there is no way to know which lied.</p>
<p>Per-epoch row:</p>

column | notes
-- | --
epoch |  
train_loss | mean over the epoch's batches
train_acc |  
val_loss | requires a held-out split — see below
val_acc |  
lr | logged even while constant; becomes the point of the column once scheduling lands
epoch_time_ms | cheap to capture, useful later when comparing against a CUDA path

<p>Output to <code>logs/run_&lt;timestamp&gt;.csv</code>. Path configurable.</p>
<h3>2. Hyperparameters must travel with the run</h3>
<p>A CSV of curves is useless for comparison if you can't tell which run was which. Each run also writes <code>logs/run_&lt;timestamp&gt;.json</code> with:</p>
<ul>
<li>architecture (layer sizes, activations)</li>
<li>initialization scheme (<code>he</code> / <code>default</code>)</li>
<li>optimizer + learning rate</li>
<li>batch size, epoch count</li>
<li>random seed</li>
</ul>
<p>Without this, the He-vs-default comparison plot is unreproducible three weeks from now.</p>
<h3>3. Use the official MNIST test set</h3>
<p>Currently train, validation, and test are all carved out of the 60k training file (shuffled, seed 42). The official 10k test file is unused.</p>
<p>This is a problem, and not a cosmetic one. <strong>MNIST's train and test sets were constructed with disjoint writers</strong> — no person's handwriting appears in both. That disjointness is the reason the official test set measures generalization to an unseen hand rather than to unseen samples from hands the model has already learned. Carving a test split out of the training file destroys that property: shuffling spreads each writer's digits across both sides of the split, so the model is evaluated on handwriting it has already seen.</p>
<p>Two consequences:</p>
<ol>
<li>The reported number is optimistically biased — it measures something easier than the benchmark does.</li>
<li>It is not comparable to any published MNIST result, all of which are against those 10k images. The README currently says "test accuracy on MNIST," which implies comparability it doesn't have.</li>
</ol>
<p>Target layout:</p>
<pre><code>train:  55k   from the 60k train file
val:     5k   from the 60k train file
test:   10k   the official t10k files — untouched until final evaluation
</code></pre>
<p>Train and val sharing writers is fine: val is for <em>choosing</em> hyperparameters, not for <em>claiming</em> performance. The claimed number gets measured on the canonical set.</p>
<p>Cost is low — the existing IDX parser points at a different filename.</p>
<p><strong>Expect the headline accuracy to drop slightly on re-measurement.</strong> Update the README to the honest number.</p>
<h3>4. Python side — <code>scripts/plot.py</code></h3>
<p>Reads one or more run CSVs, emits static PNGs to <code>assets/</code>. Matplotlib + seaborn. <strong>No Plotly</strong> — GitHub markdown strips JavaScript, so an interactive HTML plot will not render in the README and nobody will open it. Static PNGs render inline.</p>
<p>Plots:</p>
<ol>
<li><strong>Loss curves</strong> — train vs. validation, single run. The overfitting picture.</li>
<li><strong>Accuracy curves</strong> — train vs. validation, single run.</li>
<li><strong>Multi-run overlay</strong> — arbitrary N runs on one axis, labeled from the JSON sidecar. First use: He init vs. default, showing the convergence-speed delta.</li>
</ol>
<p>Style: dark background to match the repo's aesthetic, no chart junk, no gridline soup, no legend boxes with drop shadows. Two lines and an axis.</p>
<h3>5. Commit the outputs</h3>
<p>Selected PNGs go into the repo README. This is the point of the whole exercise — the plots are the evidence that the numbers in the README are real.</p>
<h2>Definition of done</h2>
<ul>
<li>[ ] <code>Logger</code> class; terminal output and CSV output share one call site</li>
<li>[ ] Per-epoch CSV written to <code>logs/</code></li>
<li>[ ] Hyperparameter JSON sidecar per run</li>
<li>[ ] Official t10k test set loaded and used as the held-out test set; val carved from the 60k train file</li>
<li>[ ] README accuracy figures re-measured against the official test set</li>
<li>[ ] <code>scripts/plot.py</code> producing the three plot types</li>
<li>[ ] He-vs-default overlay committed to <code>assets/</code> and embedded in the README</li>
</ul>
<h2>Explicitly out of scope</h2>
<ul>
<li>Interactive / web visualization → see #12 (Imperium, <code>someday</code>)</li>
<li>Per-batch (as opposed to per-epoch) logging → later, if the loss curve turns out to be too coarse to debug something specific</li>
<li>TensorBoard or any external logging dependency → the whole project is a from-scratch build; a CSV and a plotting script is the correct amount of machinery</li>
</ul></body></html>
</body>
</html>

**Comment by yash-jha-888** [comment]
Starting now for sure

**Comment by yash-jha-888** [comment]
writer's block over now ig
